File: frontend/src/services/blockfrostService.js
const BLOCKFROST_PROJECT_ID = import.meta.env.VITE_BLOCKFROST_PROJECT_ID
const BLOCKFROST_API_URL = import.meta.env.VITE_BLOCKFROST_API_URL

class BlockfrostService {
    /**
     * Check if transaction is confirmed on blockchain
     * @param {string} txHash - Transaction hash
     * @returns {Promise<string>} Status: 'confirmed', 'pending', or 'unknown'
     */
    async checkTransactionStatus(txHash) {
        try {
            const response = await fetch(`${BLOCKFROST_API_URL}/txs/${txHash}`, {
                headers: {
                    'project_id': BLOCKFROST_PROJECT_ID
                }
            })

            if (response.ok) {
                console.log(`✅ Transaction ${txHash.substring(0, 8)}... is CONFIRMED`)
                return 'confirmed'
            } else if (response.status === 404) {
                console.log(`⏳ Transaction ${txHash.substring(0, 8)}... is PENDING`)
                return 'pending'
            } else {
                console.warn(`❓ Transaction ${txHash.substring(0, 8)}... status UNKNOWN`)
                return 'unknown'
            }
        } catch (error) {
            console.error('Blockfrost API error:', error)
            return 'unknown'
        }
    }

    /**
     * Check multiple transactions at once
     * @param {Array<string>} txHashes - Array of transaction hashes
     * @returns {Promise<Array<{hash: string, status: string}>>}
     */
    async checkMultipleTransactions(txHashes) {
        const results = await Promise.all(
            txHashes.map(async (hash) => ({
                hash,
                status: await this.checkTransactionStatus(hash)
            }))
        )
        return results
    }
}

export default new BlockfrostService()

File: backend/src/main/java/csit360/g6/team/masikip/service/BlockfrostService.java
package csit360.g6.team.masikip.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class BlockfrostService {

    @Value("${blockfrost.project.id}")
    private String projectId;

    @Value("${blockfrost.api.url}")
    private String apiUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    /**
     * Check if transaction is confirmed on blockchain
     * 
     * @param txHash Transaction hash
     * @return "confirmed" if on-chain, "pending" if not found, "unknown" if error
     */
    public String checkTransactionStatus(String txHash) {
        try {
            System.out.println("🔍 Checking transaction status: " + txHash);

            String url = apiUrl + "/txs/" + txHash;

            HttpHeaders headers = new HttpHeaders();
            headers.set("project_id", projectId);

            HttpEntity<String> entity = new HttpEntity<>(headers);

            ResponseEntity<String> response = restTemplate.exchange(
                    url,
                    HttpMethod.GET,
                    entity,
                    String.class);

            // 200 OK = transaction is on-chain
            if (response.getStatusCode() == HttpStatus.OK) {
                System.out.println("✅ Transaction confirmed: " + txHash);
                return "confirmed";
            }

            return "unknown";

        } catch (HttpClientErrorException.NotFound e) {
            // 404 = transaction not yet confirmed
            System.out.println("⏳ Transaction pending: " + txHash);
            return "pending";
        } catch (Exception e) {
            System.err.println("❌ Blockfrost API error: " + e.getMessage());
            return "unknown";
        }
    }

    /**
     * Check if transaction is confirmed (boolean version)
     * 
     * @param txHash Transaction hash
     * @return true if confirmed, false otherwise
     */
    public boolean isTransactionConfirmed(String txHash) {
        return "confirmed".equals(checkTransactionStatus(txHash));
    }

    /**
     * Get transaction history for an address
     */
    public List<String> getAddressTransactions(String address) {
        try {
            String url = apiUrl + "/addresses/" + address + "/transactions?order=desc";
            HttpHeaders headers = new HttpHeaders();
            headers.set("project_id", projectId);
            HttpEntity<String> entity = new HttpEntity<>(headers);

            ResponseEntity<List> response = restTemplate.exchange(
                    url,
                    HttpMethod.GET,
                    entity,
                    List.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                List<Map<String, Object>> txs = response.getBody();
                return txs.stream()
                        .map(tx -> (String) tx.get("tx_hash"))
                        .collect(Collectors.toList());
            }
        } catch (Exception e) {
            System.err.println("❌ Failed to fetch address transactions: " + e.getMessage());
        }
        return Collections.emptyList();
    }

    /**
     * Get transaction metadata
     */
    public List<Map<String, Object>> getTransactionMetadata(String txHash) {
        try {
            String url = apiUrl + "/txs/" + txHash + "/metadata";
            HttpHeaders headers = new HttpHeaders();
            headers.set("project_id", projectId);
            HttpEntity<String> entity = new HttpEntity<>(headers);

            ResponseEntity<List> response = restTemplate.exchange(
                    url,
                    HttpMethod.GET,
                    entity,
                    List.class);

            if (response.getStatusCode() == HttpStatus.OK) {
                return response.getBody();
            }
        } catch (Exception e) {
            // 404 means no metadata, which is fine
        }
        return Collections.emptyList();
    }
}
